### PR TITLE
[EVM] Constant unfolding -Oz fallback

### PR DESCRIPTION
**Pull request by PavelKopyl**
Key changes introduced:
    
 - Transitioned from a MachineFunction pass to a Module-level pass
 - Implemented module-wide code size estimation
    (not 100% exact but sufficiently accurate for our purpose)
 - Applied transformations by iterating over functions within the module.
    For each function:
      - Handled both loop and non-loop basic blocks using the default heuristic
         (i.e., based on the presence of the OptSize attribute)
      - If the module size exceeds the defined limit, switched to a fallback mode
         that treats all functions as if they have OptSize flag.
 - Repeatedly traversed all functions, targeting only basic blocks at a specific
    loop nesting depth (starting from depth 0 for non-loop blocks), and
    re-evaluated module size. The process stops once the size drops below the
    limit, or reaching the loop depth limit.
    
This enables a balanced distribution of runtime gas regressions across functions
 within the module, resulting from aggressive constant unfolding.

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

File: llvm/test/CodeGen/EVM/constant-unfolding-Oz-fallback.ll [hunk 1]
```diff
@@ -0,0 +1,77 @@
+; RUN: llc -O3 --debug-only=evm-constant-unfolding -evm-bytecode-sizelimit=80 < %s 2>&1 | FileCheck %s
```
Comment (vladimirradosavljevic) [inline llvm/test/CodeGen/EVM/constant-unfolding-Oz-fallback.ll:1]: You will need to add here `; REQUIRES: asserts`, since this test will fail without asserts.

File: llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll [hunk 2]
```diff
@@ -60,12 +57,9 @@ define void @test4() #1 {
 ; CHECK-LABEL: test4:
 ; CHECK:       ; %bb.0: ; %entry
 ; CHECK-NEXT:    JUMPDEST
-; CHECK-NEXT:    PUSH0
-; CHECK-NEXT:    NOT
-; CHECK-NEXT:    PUSH1 0xC8
+; CHECK-NEXT:    PUSH7 0xFFFFFFFFFFFFFF
```
Comment (vladimirradosavljevic) [inline llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll:60]: IIUC, for -Oz we want to have bytecode as small as possible. Is there a reason for this change?

File: llvm/lib/Target/EVM/EVMInstrFormats.td [hunk 3]
```diff
@@ -36,32 +36,31 @@ class EVMInst<string asmstr, bit stack, int cost>
   bit IsPush = 0;
   bit StackBased = stack;
   string BaseName = NAME;
-  int GasCost     = cost;
+  bits<24> BaseGasCost = cost;
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMInstrFormats.td]: Is there are reason why we need 24 bits? We could do it in 15 bits, as CREATE and CREATE2 are 32000.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 4]
```diff
@@ -411,174 +427,313 @@ EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
...
+      TransformationCache.try_emplace({Imm, OptForSize}, std::move(*OptIt));
   assert(Res.second);
 
   return Tr;
 }
 
-bool EVMConstantUnfolding::tryUnfoldConstant(MachineInstr &MI) {
+static bool isProfitableToTranform(const APInt &Imm, const StackCostModel &Cost,
+                                   bool OptForSize) {
+  if (!OptForSize) {
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Please reverse condition and do early exit.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 5]
```diff
@@ -411,174 +427,313 @@ EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
...
+  // TODO: remove this workaround once the EVM CodeGen pipeline adopts
+  // the new pass manager, #862.
+
+  // The order of object construction is important to avoid
+  // stack-use-after-scope error.
+  MachineFunctionAnalysisManager MFAM;
+  ModuleAnalysisManager MAM;
+
+  // Register module analyses
+  MAM.registerPass([&MMI] { return MachineModuleAnalysis(MMI); });
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: This feels a little bit odd, to run passes inside a machine pass. Can we just do something like [here](https://github.com/matter-labs/era-compiler-llvm/blob/41693a3d04b33b51596613e07d14dc426ce0842d/llvm/lib/CodeGen/AsmPrinter/AsmPrinter.cpp#L1706) (to recalculate dominator tree and loop info for each function), since if I'm not mistaken under the hood these analysis will be recalculated before this pass.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 6]
```diff
@@ -411,174 +427,313 @@ EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
...
+             << LoopDepth << " ***\n";
+    });
 
-      Changed |= tryUnfoldConstant(MI);
-    }
-  }
+    MinSizeAtLoopDepthOnly = LoopDepth;
+    Changed |= processModule(ModuleCodeSize, Unfolder, MinSizeAtLoopDepthOnly,
+                             M, MMI, MFAM);
+    CodeSizeReduction = Unfolder.getCodeSizeReduction();
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Can we use some other variable instead of `CodeSizeReduction`? Statistic variables are not cheap to use, since if stats are enabled they are using atomic operations under the hood.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 7]
```diff
@@ -411,174 +427,313 @@ EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
...
+
+static bool runImpl(Module &M, ModuleAnalysisManager &AM) {
+  bool Changed = false;
+  MachineModuleInfo &MMI = AM.getResult<MachineModuleAnalysis>(M).getMMI();
+  MachineFunctionAnalysisManager &MFAM =
+      AM.getResult<MachineFunctionAnalysisManagerModuleProxy>(M).getManager();
+
+  const auto &EVMTM = static_cast<const EVMTargetMachine &>(MMI.getTarget());
+  ConstantUnfolder Unfolder(&M.getContext(),
+                            EVMTM.getSubtarget()->getInstrInfo());
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Usually in module passes, TTI is retrieved for each function like this `MF.getSubtarget<EVMSubtarget>().getInstrInfo()`.
Could we do the same?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 8]
```diff
@@ -24,209 +24,243 @@
...
-                                   cl::desc("Gas weight used as a factor in the"
-                                            "transformation cost function"));
-
 static cl::opt<unsigned>
     LoopDepthLimit("evm-const-loop-depth-limit", cl::Hidden, cl::init(2),
                    cl::desc("The maximum loop depth at which constant"
                             "unfolding is still considered beneficial"));
 
+static cl::opt<unsigned>
+    CodeSizeLimit("evm-bytecode-sizelimit", cl::Hidden, cl::init(24576),
```
Comment (vladimirradosavljevic) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:71]: This is true for runtime code, but for init code, limit is 48 KB. I think FE should set this accordingly if it is runtime or init code.

**Review by vladimirradosavljevic — DISMISSED** [review]
I would suggest to do traverse over functions only once, and to collect everything you need to do transformation:
1. Move code size calculation in unfold pass, and during this, also collect PUSHes that we will transform. You can distinguish between PUSHes that will be transformed regardless of bytecode size and others if we are exceeding the limit.
2. With this approach, you won't need MLI analysis after you finish with that function.
3. In case we could end up with minsize functions during optimization for speed, you will need to distinguish between these two transformations. In case we don't, we can create a TODO.

**Comment by PavelKopyl** [comment]
> I would suggest to do traverse over functions only once, and to collect everything you need to do transformation:
> 
> 1. Move code size calculation in unfold pass, and during this, also collect PUSHes that we will transform. You can distinguish between PUSHes that will be transformed regardless of bytecode size and others if we are exceeding the limit.
> 2. With this approach, you won't need MLI analysis after you finish with that function.
> 3. In case we could end up with minsize functions during optimization for speed, you will need to distinguish between these two transformations. In case we don't, we can create a TODO.

Thank you for the review!
One note regarding caching of instructions: the idea behind this check was to perform unfolding incrementally. Initially, it targets all blocks in OptForSpeed mode. Later, in OptForSize mode, I planned to expand it in stages, first for non-loop basic blocks, then for loops at depth 1, followed by loops at depth 2, and so on. Given this staged approach, maintaining a large number of caches would be both complex and potentially inefficient.

File: llvm/lib/Target/EVM/EVMInstrFormats.td (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMInstrFormats.td]: You are right. I allocated 3 bytes just to be prepared in case  of changes in base gas fees, but OK I'll shrink it to 2 bytes.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 5 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Yes, this is a bit unconventional, I implemented it to cache loop analysis results, since we iterate over the machine function's basic blocks multiple times. Without caching, I would have to recompute this manually each time. It's also a step toward adopting the new pass manager for CodeGen.

**Comment by vladimirradosavljevic** [comment]
> > I would suggest to do traverse over functions only once, and to collect everything you need to do transformation:
> > 
> > 1. Move code size calculation in unfold pass, and during this, also collect PUSHes that we will transform. You can distinguish between PUSHes that will be transformed regardless of bytecode size and others if we are exceeding the limit.
> > 2. With this approach, you won't need MLI analysis after you finish with that function.
> > 3. In case we could end up with minsize functions during optimization for speed, you will need to distinguish between these two transformations. In case we don't, we can create a TODO.
> 
> Thank you for the review! One note regarding caching of instructions: the idea behind this check was to perform unfolding incrementally. Initially, it targets all blocks in OptForSpeed mode. Later, in OptForSize mode, I planned to expand it in stages, first for non-loop basic blocks, then for loops at depth 1, followed by loops at depth 2, and so on. Given this staged approach, maintaining a large number of caches would be both complex and potentially inefficient.

IIRC, this PR increases compilation time, so traversing over functions only once should fix that.
Since you're already doing staging, you could maintain separate vectors for each stage and traverse them accordingly. Functionally, this isn’t much different from the current approach of manually switching stages, or I'm missing something?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 5 above)
Comment (vladimirradosavljevic) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: We could cache analysis and reuse them when needed, if we are going to traverse over functions more than once.
Could you please explain how it will be different with the new pass manager for Module passes?

File: llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+#include "MCTargetDesc/EVMMCTargetDesc.h"
+#include "llvm/CodeGen/MachineModuleInfo.h"
+#include "llvm/CodeGen/TargetInstrInfo.h"
+#include "llvm/CodeGen/TargetSubtargetInfo.h"
+#include "llvm/IR/Module.h"
+
+using namespace llvm;
+
+// This is the copied from AsmPrinter::isBlockOnlyReachableByFallthrough.
+static bool isBlockOnlyReachableByFallthrough(const MachineBasicBlock *MBB) {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp:25]: Just to check: could't `analyzeBranch()` or `MachineBasicBlock::getFallThrough()` help here?

File: llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+    Size += std::accumulate(MBB.begin(), MBB.end(), 0,
+                            [&TII](unsigned Sum, const MachineInstr &MI) {
+                              return Sum + getInstSize(MI, TII);
+                            });
+  }
+  return Size;
+}
+
+uint64_t llvm::EVM::calculateModuleCodeSize(Module &M,
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp:126]: Dow we want a unit test for this?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 11]
```diff
@@ -24,209 +24,243 @@
...
-                               "required for constant unfolding when"
-                               "optimizing for speed"));
+    SizeReductionThreshold("evm-const-unfolding-size-reduction-threshold",
+                           cl::Hidden, cl::init(2.5),
+                           cl::desc("Minimum original-to-unfolded size ratio"
+                                    "required for constant unfolding when"
+                                    "optimizing for speed"));
 
 static cl::opt<unsigned>
     InstrNumLimitUnfolInto("evm-const-unfolding-inst-num-limit", cl::Hidden,
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: typo: InstrNumLimitUnfol**d**Into

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 12]
```diff
@@ -24,209 +24,243 @@
...
-    void insertInstr(MachineInstr &MI, unsigned Opcode) {
-      BuildMI(*MI.getParent(), MI, MI.getDebugLoc(), TII->get(Opcode));
-    }
+private:
+  LLVMContext *Context{};
+  const TargetInstrInfo *TII{};
+
+  // The 'second' field can be set to 0 or 1, indicating whether to
+  // optimize for performance or size.
+  using TransformationKey = std::pair<APInt, int>;
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:182]: Do we need `int` if the second field can be only `0` or `1`?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 13]
```diff
@@ -24,209 +24,243 @@
...
 
-FunctionPass *llvm::createEVMConstantUnfolding() {
+ModulePass *llvm::createEVMConstantUnfolding() {
   return new EVMConstantUnfolding();
 }
 
-const EVMConstantUnfolding::TransformationBuilder *
-EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
-  if (auto It = TransformationCache.find(Imm);
+static bool isBetterCandidate(const TransformationBuilder &A,
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I'm thinking - could this be a `operator<` while renaming TransformationBuilder to TransformationCandidate? It will simplify minelement call.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 8 above)
Comment (akiramenai) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:71]: I think it should be module level metadata actually. A FE (think not just solx FE!) **must** specify what kind of module it is as limitations are different. Also execution profile is different too so we might want to make some alterations in optimizations as well.
If nothing is specified we can assume runtime but emit a warning. CC: @hedgar2017.

**Review by akiramenai — COMMENTED** [review]
The overall idea look good to me. I have stylistic remarks only.
I still need to think about running passes inside a pass to suggest an alternative. It's better to be resolved.

File: llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll (hunk 2 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/EVM/constant-unfolding-Oz.ll:60]: I've changed the transformation weight function to
```
static bool isBetterCandidate(const TransformationBuilder &A,
                                const TransformationBuilder &B, bool OptForSize) {
    static constexpr unsigned Weight = 2;
    const StackCostModel &CostA = A.getCost();
    const StackCostModel &CostB = B.getCost();
    unsigned ScoreA = 0;
    unsigned ScoreB = 0;
  
    if (OptForSize) {
      ScoreA = (CostA.getByteSize() * Weight) + CostA.getGas();
      ScoreB = (CostB.getByteSize() * Weight) + CostB.getGas();
    } else {
      ScoreA = CostA.getByteSize() + (Weight * CostA.getGas());
      ScoreB = CostB.getByteSize() + (Weight * CostB.getGas());
    }
    if (ScoreA != ScoreB)
      return ScoreA < ScoreB;
  
    return CostA.getInstrCount() < CostB.getInstrCount();
  }
```
It seems reasonable to consider the gas cost even in -Oz mode, since two transformations might differ by just one byte in size but have significantly different gas usage. Probably I'm wrong, so I'd like to compare the effect on Foundry projects.

**Comment by PavelKopyl** [comment]
> > > I would suggest to do traverse over functions only once, and to collect everything you need to do transformation:
> > > 
> > > 1. Move code size calculation in unfold pass, and during this, also collect PUSHes that we will transform. You can distinguish between PUSHes that will be transformed regardless of bytecode size and others if we are exceeding the limit.
> > > 2. With this approach, you won't need MLI analysis after you finish with that function.
> > > 3. In case we could end up with minsize functions during optimization for speed, you will need to distinguish between these two transformations. In case we don't, we can create a TODO.
> > 
> > 
> > Thank you for the review! One note regarding caching of instructions: the idea behind this check was to perform unfolding incrementally. Initially, it targets all blocks in OptForSpeed mode. Later, in OptForSize mode, I planned to expand it in stages, first for non-loop basic blocks, then for loops at depth 1, followed by loops at depth 2, and so on. Given this staged approach, maintaining a large number of caches would be both complex and potentially inefficient.
> 
> IIRC, this PR increases compilation time, so traversing over functions only once should fix that. Since you're already doing staging, you could maintain separate vectors for each stage and traverse them accordingly. Functionally, this isn’t much different from the current approach of manually switching stages, or I'm missing something?

I agree, will do this, thank you.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 12 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:182]: In this case, I'll need to provide a custom caching function for those keys, whereas a default implementation already exists for `int`.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 5 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: >We could cache analysis and reuse them when needed, if we are going to traverse over functions more than once.
Yeap, but this has to be done manually. 
>Could you please explain how it will be different with the new pass manager for Module passes?
In this case, it's enough to do following:
```
static bool runImpl(Module &M, ModuleAnalysisManager &AM) {
    MachineFunctionAnalysisManager &MFAM =
        AM.getResult<MachineFunctionAnalysisManagerModuleProxy>(M).getManager();
    MachineLoopInfo &MLI = MFAM.getResult<MachineLoopAnalysis>(*MF);
    ...
   }
```
and the new PM will run other analyses and cache their results. As I know, we cannot do this (i.e., access MF analyses from a Module pass level) with the legacy PM. Once we switch to the new PM, I'll just remove all the code from `runOnModule(Module &M)` that creates new PM instance, and that's it. In case of a manual caching, I'll need to do more deep clean up.

File: llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp (hunk 9 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp:25]: I think this is the simplest implementation.
MachineBasicBlock::getFallThrough() returns the block that a given MBB falls through to, whereas here we need to solve the reverse problem.

File: llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp (hunk 10 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp:126]: Sure, added: llvm/unittests/Target/EVM/BytecodeSize.cpp

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 8 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp:71]: I agree, but I'd make it as a separate task. For this specific pass it's enough just to know the size limit. Is that OK?

**Comment by PavelKopyl** [comment]
> The overall idea look good to me. I have stylistic remarks only. I still need to think about running passes inside a pass to suggest an alternative. It's better to be resolved.

I removed the trick regarding new pass manager, so now I manually calculate MachineLoopInfo.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 13 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: I've renamed TransformationBuilder -> TransformationCandidate, yes this name looks better. But in the comparator I need third argument - OptForSize, so I don't think introducing `operator<` will simplify the code.

File: llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    break;
+  case EVM::PseudoCALL:
+    // In case that function call has a return label, we will emit JUMPDEST,
+    // so take it into account.
+    if (MI.getNumExplicitOperands() > 1)
+      Size = TII->get(EVM::JUMPDEST_S).getSize();
+    LLVM_FALLTHROUGH;
+  case EVM::PseudoJUMPI:
+  case EVM::PseudoJUMP:
+    // Even though we are emitting PUSH4_S, most probably this will be relaxed
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMCalculateModuleSize.cpp]: // We emit PUSH4_S here. The linker usually relaxes it to PUSH2_S,
// since a 16-bit immediate covers the 24,576-byte EVM runtime code cap (EIP-170).
// If a wider immediate were ever required, the contract already exceeds the cap,
// so the push width is moot.

File: llvm/lib/Target/EVM/EVMTargetMachine.h [hunk 15]
```diff
@@ -37,6 +37,8 @@ class EVMTargetMachine final : public LLVMTargetMachine {
     return &Subtarget;
   }
 
+  const EVMSubtarget *getSubtarget() const { return &Subtarget; }
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMTargetMachine.h]: Why is it needed?

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp [hunk 16]
```diff
@@ -24,209 +24,245 @@
...
+// allowing the TransformationCandidate object to be reused for repeated
+// constants. As new instructions are added, the StackCostModel is used
+// to track the accumulated cost.
+class TransformationCandidate {
+public:
+  TransformationCandidate(LLVMContext &Context, const TargetInstrInfo *TII)
+      : Context(Context), TII(TII) {}
 
-    void add() { accountInstr(EVMCOST::ADD); }
+  void addSub() {
```
Comment (akiramenai) [inline llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Could you check if it's covered by regression / unit tests?
My coverage report complaints about this function.

**Review by akiramenai — APPROVED** [review]
LGTM

File: llvm/lib/Target/EVM/EVMTargetMachine.h (hunk 15 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMTargetMachine.h]: Thanks, missed to remove.

File: llvm/lib/Target/EVM/EVMConstantUnfolding.cpp (hunk 16 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: This also needs to be removed.

Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMConstantUnfolding.cpp]: Done.

File: llvm/lib/Target/EVM/EVMTargetMachine.h (hunk 15 above)
Comment (PavelKopyl) [reply llvm/lib/Target/EVM/EVMTargetMachine.h]: Done.
